upgrades/upgrade_system.gd

```diff
@@ -0,0 +1,31 @@
+class_name UpgradeSystem
+extends Resource
+
+var _upgrades = UpgradeDefinition.get_all()
+
+func _create_upgrade_levels(upgrades)->Dictionary:
+	var result: Dictionary = {}
+	for upgrade in upgrades:
+		result[upgrade.id] = 0
+	return result
+
+var _upgrade_levels: Dictionary
+
+var _state: GameState
+
+func _init(state) -> void:
+	_upgrade_levels = _create_upgrade_levels(_upgrades)
+	_state = state
+
+	SignalBus.attempt_upgrade.connect(_handle_attempt_upgrade)
+
+func can_afford(upgrade: UpgradeDefinition) -> bool:
+	return _state.get_coins() >= upgrade.cost
+
+func _handle_attempt_upgrade(upgrade: UpgradeDefinition) -> void:
+	if not can_afford(upgrade): return
+
+	_state.decrease_coins.emit(upgrade.cost)
+	_upgrade_levels[upgrade.id] += 1
+
```

Comment:
would this be where we'd adjust the cost of the upgrade?

we could have an Upgrade database that's just a dictionary

```gdscript
var Upgrades = {
  "clip-size": {
    max_level: 13,
    get_cost: func(level: int, base_cost) -> int:
      # where 10 is pretty much the base cost but doesn't really need to be defined anywhere since
      # we're just establishing a formula which could be linear, log, exp, linear, or whatever to ensure the cost
     # increases with the level
      return 10 + ceil(level * 1.5)
  }
}
```